// unfolding-word/uw-generate-usfm.js
/**
 * This script is designed to parse usfm files, and generate the appropriate HTML for each chapter of the books of the Bible.
 * The HTML is based on what is appropriate for this app.  To use this generator, add `uw_usfm` in the generator section
 * of your info.json file.
 *
 * @author Johnathan Pulos <johnathan@missionaldigerati.org>
 */
var uwGenerateUsfm = function() {
  /**
   * This classes main object
   *
   * @type {Object}
   * @access private
   */
  var uwObject = {};
  /**
   * Nodejs package filesystem for writing files
   *
   * @type {Object}
   * @access private
   */
  var fileSystem = require('fs');
  /**
   * Nodejs package path for generating correct pathing
   *
   * @type {Object}
   * @access private
   */
  var path = require('path');
  /**
   * A cusom NodeJS file for parsing the data in USFM format
   *
   * @type {Object}
   * @access private
   */
  var usfmParser = require('usfm_parser');
  /**
   * A cusom NodeJS file for parsing Bible data
   *
   * @type {Object}
   * @access private
   */
  var bibleDataParser = require('bible_data');
  /**
   * A cusom NodeJS file for formatting Bible data
   *
   * @type {Object}
   * @access private
   */
  var bibleFormatter = require('bible_formatter');

  /**
   * The base path to the folder containing the USFM files
   *
   * @type {String}
   * @access private
   */
  var inputBasePath = '';
  /**
   * A JSON object storing the return data for the generate() function
   *
   * @type {Object}
   * @access private
   */
  var bibleData = {};
  /**
   * The main method for generating the appropriate HTML for each USFM tag.  This method does the following:
   * 1) Iterates over all the files (Note:: Each file holds one book of the Bible) in inputBasePath that has a .usfm extension
   * 2) Iterates over each line in the file
   * 3) Parses the USFM file and generates the appropriate HTML for each chapter in the book
   * 4) Return an JSON Object with various information about the book including the chapter html
   * 
   * In the Chapter data object, each chapter contains the folowing data:
   * 
   *  { id: 'RV21',
   *    html: 'HTML FOR THE CHAPTER',
   *    notes: '',
   *    previd: 'RV20',
   *    nextid: 'RV22' 
   *  }
   *
   * @param  {String}   inputBasePath   The path to the folder containing the USFM files
   * @param  {Object}   info            The JSON object of information in the info.json file in the inputBasePath
   * @param  {Boolean}  createIndex     Do you want to create an index of the verses? (See verse_indexer.js)    
   * @param  {callback} startProgress   The callback to start the progress bar
   * @param  {callback} updateProgress  The callback for updating the current progress
   *
   * @return {Object}                   A JSON object containing chapterData, indexData, indexLemmaData, and aboutHtml
   *
   * @author Johnathan Pulos <johnathan@missionaldigerati.org>
   */
  uwObject.generate = function(_inputBasePath, info, createIndex, startProgress, updateProgress) {
    inputBasePath = _inputBasePath;
    /**
     * This is the returned object of this method.
     *
     * @type {Object}
     */
    bibleData = {
      chapterData:    [],
      indexData:      {},
      indexLemmaData: {},
      aboutHtml:      ''
    };
    var usfmFiles = fileSystem.readdirSync(inputBasePath);

    bibleData.aboutHtml = getAboutContent();
    startProgress(usfmFiles.length, 'Books');

    usfmFiles.forEach(function(filename) {
      /**
       * Only work with USFM files
       */
      if (filename.indexOf('.usfm') == -1) {
        return;
      }
      var fileContent = fileSystem.readFileSync(path.join(inputBasePath, filename), 'utf8');
      var lines = fileContent.split('\n');

      /**
       * Set various variables used in the loop
       */
      var currentBookInfo = null;
      var currentChapterCode = null;
      var currentChapterNumber = 0;
      for (var i = 0, il = lines.length; i < il; i++) {
        var line = lines[i];
        var usfmData = usfmParser.parseLine(line);

        if (usfmData == null) {
          /**
           * We have no data for the line
           */
          return;
        };

        switch (usfmData.key) {
          case 'ide':
            /**
             * We are assuming all files are UTF-8 encoded
             */
          break;
          case 'id':
            var bookId = usfmData.text.split(' ')[0].trim().toUpperCase();
            currentBookInfo = bibleDataParser.getBookInfoByUsfmCode(bookId);
          break;
          case 'c':
            /**
             * The chapter number is passed in the usfmData.text string
             * ex. { key: 'c', number: '', text: '1' }
             */
            currentChapterNumber = parseInt(usfmData.text);
            if (currentChapterNumber > 1) {
              /**
               * We have a new chapter
               */
              closeChapter(currentChapterCode);
            };
            currentChapterCode = bibleFormatter.formatChapterCode(currentBookInfo.dbsCode, currentChapterNumber);
          break;
        };
      };
      /**
       * End of iterating over each line of the file
       * Now to clean up what remains
       */
       closeChapter(currentChapterCode);
    });
    /**
     * End of iterating over each file
     */

    /**
     * Return the Bible data
     */
    return bibleData;
  };
  /**
   * Private methods
   */
  /**
   * Check whether there is an about.html file in the inputBasePath, if so return it's contents.
   * If not, then return an empty string.
   *
   * @return {String} An HTML string containing information about the Bible
   *
   * @author Johnathan Pulos <johnathan@missionaldigerati.org>
   */
  function getAboutContent() {
    var aboutPath = path.join(inputBasePath, 'about.html');
    if (fileSystem.existsSync(aboutPath)) {
      return fileSystem.readFileSync(aboutPath, 'utf8');
    }
    return '';
  };
  /**
   * Closes and stores the chapter data in the bibleData.chapterData object.
   *
   * @param  {String} id A combination of the book and chapter id (ex. JM1)
   *
   * @return {void}
   * @throws {Error} If bibleData is missing the chapterData property
   *
   * @author Johnathan Pulos <johnathan@missionaldigerati.org>
   */
  function closeChapter(id) {
    var chapterData = {
      id: id
    };
    if (bibleData.hasOwnProperty('chapterData')) {
      bibleData.chapterData.push(chapterData);
    } else {
      throw new Error('bibleData is missing the property chapterData');
    }
  };
  /**
   * Return this object
   */
  return uwObject;
};
/**
 * Expose the library
 *
 */
uw = new uwGenerateUsfm;
exports = module.exports = uw;
